Compilation error - Init from incompatible pointer type

Hello!

So, I'm getting the following error when I try and `make` on Fedora 25 (`uname -a` data is Linux hostname 4.13.5-100.fc25.x86_64 #1 SMP Fri Oct 6 01:01:44 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux - in case it is of use)

```/intrepid-socketcan-kernel-module
$>make 
make -C /lib/modules/4.13.5-100.fc25.x86_64/build M=/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module modules
make[1]: Entering directory '/usr/src/kernels/4.13.5-100.fc25.x86_64'
Makefile:940: "Cannot use CONFIG_STACK_VALIDATION, please install libelf-dev, libelf-devel or elfutils-libelf-devel"
  CC [M]  /home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module/intrepid.o
/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module/intrepid.c:501:27: error: initialization from incompatible pointer type [-Werror=incompatible-pointer-types]
         .fault          = intrepid_vm_fault
                           ^~~~~~~~~~~~~~~~~
/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module/intrepid.c:501:27: note: (near initialization for ‘intrepid_vm_ops.fault’)
cc1: some warnings being treated as errors
scripts/Makefile.build:308: recipe for target '/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module/intrepid.o' failed
make[2]: *** [/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module/intrepid.o] Error 1
Makefile:1516: recipe for target '_module_/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module' failed
make[1]: *** [_module_/home/mc/srlabs/research/ECU/intrepid-socketcan-kernel-module] Error 2
make[1]: Leaving directory '/usr/src/kernels/4.13.5-100.fc25.x86_64'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2```

**NB** - I've already edited the file to replace the destructors with `priv_destructor` at GCC's suggestion (it should be fine, I would think?) 

It seems that `.fault` (based off what I read in `/include/linux/mm.h`) needs it to be a `vm_fault` struct? but I'm not an expert on writing linux drivers, I'm afraid :P 

Many thanks for your time, M.

--- a/intrepid.c
+++ b/intrepid.c
@@ -47,6 +47,7 @@
 #include <linux/can/error.h>
 #include <linux/wait.h>
 #include <linux/poll.h>
+#include <linux/version.h>
 
 #define KO_DESC "Netdevice driver for Intrepid CAN/Ethernet devices"
 #define KO_VERSION "1.0"
@@ -301,7 +302,12 @@ static int intrepid_add_if(struct intrepid_netdevice** result)
         dev->base_addr          = i;
         dev->flags             |= IFF_ECHO;
         dev->netdev_ops         = &intrepid_netdevice_ops;
-        dev->destructor         = intrepid_netdevice_free;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,11,9)
+        dev->destructor    = intrepid_netdevice_free;
+#else
+        dev->needs_free_netdev  = true;
+        dev->priv_destructor    = intrepid_netdevice_free;
+#endif
         ics                     = netdev_priv(dev);
         ics->dev                = dev;
         ics->is_stopped         = 0;
@@ -488,8 +494,15 @@ static long intrepid_dev_ioctl(struct file *fp, unsigned int cmd, unsigned long 
 }
 
 /* when the mmap()ed pages are first accesed by usermode there will be a page fault.
- * here we simply linerally map in the big vmalloc() we got */
-static int intrepid_vm_fault(struct vm_area_struct *vma, struct vm_fault *vmf)
+ * here we simply linerally map in the big vmalloc() we got.
+ *
+ * Starting in kernel version 4.11, (struct vm_operations_struct *)->fault() no
+ * longer takes the vma parameter (since it resides in vmf) */
+static int intrepid_vm_fault(
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,11,0)
+        struct vm_area_struct *vma,
+#endif
+        struct vm_fault *vmf)
 {
         vmf->page = vmalloc_to_page(shared_mem + (vmf->pgoff << PAGE_SHIFT));
         get_page(vmf->page); /* increment reference count, very important */
@@ -721,7 +734,13 @@ static __exit void intrepid_exit(void)
         {
                 if (net_devices[i] != NULL)
                 {
+                    // Kernel version 4.11.9 changed the destructor function to
+                    // priv_destructor
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,11,9)
                         net_devices[i]->destructor = NULL; /* no dangling callbacks */
+#else
+                        net_devices[i]->priv_destructor = NULL;
+#endif
                         intrepid_remove_if(i);
                 }
         }

